### ISSUE TITLE
Error with Nested Models

### ISSUE THREAD

**Issue by Abdiflame** [Dobiasd/frugally-deep#233]
Hello @Dobiasd, once again _**congratulations**_ for your awesome work, it's helping me a lot with my studies and projects :)

**My situation is as follows:**
I can successfully load/predict simple models, CNN models, and pre-trained VGG with fdeep library. 
However, I am having a big issue when trying to load a "nested model".

My model its an **LRCN** kind (Long-term Recurrent Convolutional Networks)
It's composed of a time distributed VGG pre-trained model + LSTM layers.

(obs: sorry about screenshots, but I believe that is easier to explain)

![model_1](https://user-images.githubusercontent.com/5551460/87737364-b719a100-c815-11ea-9484-ec609e68920b.PNG)
![model_1b](https://user-images.githubusercontent.com/5551460/87737374-bc76eb80-c815-11ea-8fa9-82a1137c00a2.PNG)

And then, after transforming the .h5 file to .json file through "convert_model.py" and applying it to the C++ project.
I got the following error/situation:

![issue](https://user-images.githubusercontent.com/5551460/87737871-11673180-c817-11ea-95b0-90fa41ab63e1.PNG)

The model .h5 file can be found here:

https://drive.google.com/file/d/1yCSLSYPIkGWb8vuwDcixrqsOle27DfX8/view?usp=sharing

One interesting thing that I've found is that, if I change the LRCN architecture to this one:

![model_2](https://user-images.githubusercontent.com/5551460/87738076-8cc8e300-c817-11ea-8302-3ad991745fa0.PNG)
![model_2b](https://user-images.githubusercontent.com/5551460/87738084-8f2b3d00-c817-11ea-87d5-3436b4efd7d0.PNG)

The model is successfully loaded by fdeep on my C++ project. However, I am not sure if it's possible to use the pre-trained weights for training/prediction and if this is the right way to build this model.

Thanks in advance, looking forward to hearing from you.
👍

**Comment by Dobiasd** [comment]
Thanks for the very good report. I'll look into it next week and get back to you with the results.

**Comment by Dobiasd** [comment]
Status update: I've reproduced your issue locally. Here is the `Dockerfile` I used: https://gist.github.com/Dobiasd/b7883c7875f6a131af47e515f13193d3

Output:

```
Loading json ... done. elapsed time: 0.234604 s
Building model ... terminate called after throwing an instance of 'std::runtime_error'
  what():  invalid float array format
Aborted (core dumped)
```

I'll investigate and get back to you with the results.

**Comment by Dobiasd** [comment]
Ok, the problem is, frugally-deep currently does not support using full models as the inner layer of `TimeDistributed`. But it would be nice if it would. :slightly_smiling_face:

So I'm now working on adding support for it [here](https://github.com/Dobiasd/frugally-deep/pull/234).

**Comment by Abdiflame** [comment]
> Ok, the problem is, frugally-deep currently does not support using full models as the inner layer of `TimeDistributed`. But it would be nice if it would. 🙂
> 
> So I'm now working on adding support for it [here](https://github.com/Dobiasd/frugally-deep/pull/234). (I've re-created the issue in one of the test models.)

I see, yes you are right. Would be very convenient to be able to use full models as the inner layer of `TimeDistributed`.

Thanks for your consideration and hard-working. Meanwhile, I will keep trying to overcome this problem by expanding the inner layers (VGG) as I showed above.
Thanks 👍

**Comment by Abdiflame** [comment]
I would like to ask one extra question if it's not inconvenient for you.

I am having some troubles to generate the input for `model.predict({ **sequence_input** }) `  as a sequence of images.
E.g. the input shape on python is _(1, 8, 159, 224, 3)_.
By now: `const fdeep::tensor sequence_input(fdeep::tensor_shape(8, 159, 224, 3), 0.0f);` input seems to work.
However, I wanna use my sequence of images instead.

Following the FAQ, we can find the `fdeep::tensor_from_bytes()` function used for one single OPENCV image. But how can we make a sequence of images as one tensor input? 

Thanks in advance :)

**Comment by Dobiasd** [comment]
So, if I understand correctly, the question basically boils down on how to concatenate tensors in the "direction" of their fourth dimension.

The single tensors, that need to be concatenated, can be created by multiple calls to `fdeep::tensor_from_bytes` (one per image), and of course all have the same shape.

The following minimal code example shows two possible ways to do this. Please let me know if this helps. :slightly_smiling_face:

```cpp
#include "fdeep/fdeep.hpp"

int main()
{
    // Might be the result of multiple calls to fdeep::tensor_from_bytes.
    const std::vector<fdeep::tensor> tensors = {
        fdeep::tensor(fdeep::tensor_shape(159, 224, 3), 0),
        fdeep::tensor(fdeep::tensor_shape(159, 224, 3), 0),
        fdeep::tensor(fdeep::tensor_shape(159, 224, 3), 0),
        fdeep::tensor(fdeep::tensor_shape(159, 224, 3), 0),
        fdeep::tensor(fdeep::tensor_shape(159, 224, 3), 0),
        fdeep::tensor(fdeep::tensor_shape(159, 224, 3), 0),
        fdeep::tensor(fdeep::tensor_shape(159, 224, 3), 0),
        fdeep::tensor(fdeep::tensor_shape(159, 224, 3), 0)
    };
    // Now, fill values by concatenating tensors.

    // Way 1: Iterating the dimensions explicitly.
    fdeep::tensor tensor1 = fdeep::tensor(fdeep::tensor_shape(tensors.size(), 159, 224, 3), 0);
    for (std::size_t dim4 = 0; dim4 < tensor1.shape().size_dim_4_; ++dim4)
    {
        for (std::size_t y = 0; y < tensor1.shape().height_; ++y)
        {
            for (std::size_t x = 0; x < tensor1.shape().width_; ++x)
            {
                for (std::size_t z = 0; z < tensor1.shape().depth_; ++z)
                {
                    tensor1.set_ignore_rank(fdeep::tensor_pos(0, dim4, y, x, z),
                        tensors[dim4].get_ignore_rank(fdeep::tensor_pos(0, 0, y, x, z)));
                }
            }
        }
    }

    // Way 2: Getting the values as vector and concatenating them.
    const auto values = fplus::transform_and_concat([](const auto& t) {
        return t.to_vector();
    }, tensors);
    fdeep::tensor tensor2 = fdeep::tensor(fdeep::tensor_shape(tensors.size(), 159, 224, 3), values);
}
```

**Comment by Abdiflame** [comment]
Yes, it helped a lot, thanks! 💯 

I tried here and worked in this way.
However, unfortunately, the `model.prediction()` speed is still very slow! :(

![Capture](https://user-images.githubusercontent.com/5551460/88148368-736dcf80-cc39-11ea-85e4-01cf0cc438b0.PNG)


It's taking around ~7 seconds for the "Running Test" and same ~7 seconds the `model.prediction()`

**Comment by Dobiasd** [comment]
Could you please upload this model too (also in `h5`) format, so I can test the performance here?

**Comment by Abdiflame** [comment]
> Could you please upload this model too (also in `h5`) format, so I can test the performance here?

Sure, here it is:
https://drive.google.com/file/d/1bYVdRXKQ4gl8f6ms8TqEu9wFJl5HYbjf/view?usp=sharing

**Comment by Dobiasd** [comment]
Just tested on my machine ([`Dockerfile`](https://gist.github.com/Dobiasd/d9d62136613419bb6eeb43388406b7b2)), and I get ~ `2 s` for TensorFlow (no GPU, single CPU) and ~ `8 s` with frugally-deep. So, yeah, TensorFlow currently is about 4 times faster on that model architecture compared to frugally-deep (at least on my PC). I don't have a quick fix for that at hand. Sorry about that. The only possible consolation I can provide is, that if for your use-case throughput if more important than latency, you might be able to improve it by [running multiple predictions in parallel](https://github.com/Dobiasd/frugally-deep/blob/master/FAQ.md#does-frugally-deep-support-multiple-cpus).

**Comment by Abdiflame** [comment]
> Just tested on my machine ([`Dockerfile`](https://gist.github.com/Dobiasd/d9d62136613419bb6eeb43388406b7b2)), and I get ~ `2 s` for TensorFlow (no GPU, single CPU) and ~ `8 s` with frugally-deep. So, yeah, TensorFlow currently is about 4 times faster on that model architecture compared to frugally-deep (at least on my PC). I don't have a quick fix for that at hand. Sorry about that. The only possible consolation I can provide is, that if for your use-case throughput if more important than latency, you might be able to improve it by [running multiple predictions in parallel](https://github.com/Dobiasd/frugally-deep/blob/master/FAQ.md#does-frugally-deep-support-multiple-cpus).

Alright, thank you very much for your kindly reply.

I believe that I have no option nor try to find another way to implement the model on C++ or change the architecture without decreasing the accuracy. The speed is very important to my task :( 

Meanwhile, I will be waiting for any updates from the frugally-deep.
Thanks in advance, have a great week 👍

**Comment by Dobiasd** [comment]
@Abdiflame Sorry for the delay. I'm working on it right now, but it seems quite complex to solve. Is this issue still relevant to you, or are you using a different solution now?

**Comment by Abdiflame** [comment]
> @Abdiflame Sorry for the delay. I'm working on it ([pull request 234](https://github.com/Dobiasd/frugally-deep/pull/234)) today, but it seems quite complex to solve. Is this issue still relevant to you, or are you using a different solution now?

Hello @Dobiasd , no worries about the late reply :)
Yea, actually I am using python embedding inside a C++ code project and running a separate python environment for now.
Would be really nice to use _fdeep_ if these nested/speed problems were fixed. But don't worry too much about it ^^.

Thanks again for your collaboration 👍

**Comment by Dobiasd** [comment]
@Abdiflame Supposing `fdeep` would support nested models in `TimeDistributed` layers (without improving the performance, i.e., speed), would it be useful to you, or would the performance difference (compared to TF) still be too big for your use-case?

**Comment by Dobiasd** [comment]
@Abdiflame `*ping*` :slightly_smiling_face:

> Supposing `fdeep` would support nested models in `TimeDistributed` layers (without improving the performance, i.e., speed), would it be useful to you, or would the performance difference (compared to TF) still be too big for your use-case?

In case it would not help anyway, I'd simply close [pull/234](https://github.com/Dobiasd/frugally-deep/pull/234).
